diff -Naur miniupnpd-bsp32.a/minissdp.c miniupnpd-bsp32.b/minissdp.c
--- miniupnpd-bsp32.a/minissdp.c	2011-06-28 16:41:16.230924970 +0300
+++ miniupnpd-bsp32.b/minissdp.c	2011-06-29 09:20:58.146424501 +0300
@@ -497,12 +497,14 @@
 	                 "NOTIFY * HTTP/1.1\r\n"
 	                 "HOST:%s:%d\r\n"
 	                 "NT:%s%s\r\n"
-	                 "USN:%s::%s%s\r\n"
+	                 "USN:%s%s%s%s\r\n"
 	                 "NTS:ssdp:byebye\r\n"
 	                 "\r\n",
 	                 SSDP_MCAST_ADDR, SSDP_PORT,
-					 known_service_types[i], (i==0?"":"1"),
-	                 uuidvalue, known_service_types[i], (i==0?"":"1"));
+	                 known_service_types[i], ((i==0 || i==1 || i==3 || i==5)?"":"1"),
+	                 known_service_uuids[i], (i==1 || i==3 || i==5)?"":"::",
+	                 (i==1 || i==3 || i==5)?"":known_service_types[i],
+	                 ((i==0 || i==1 || i==3 || i==5)?"":"1"));
 	        n = sendto(sockets[j], bufr, l, 0,
 	                   (struct sockaddr *)&sockname, sizeof(struct sockaddr_in) );
 			if(n < 0)
