--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -67,16 +67,16 @@ all: $(TARGETS)
	@echo "Type 'make install' as root to install the software."

 pppoe-sniff: pppoe-sniff.o if.o common.o debug.o
-	@CC@ -o pppoe-sniff pppoe-sniff.o if.o common.o debug.o
+	@CC@ $(LDFLAGS) -o pppoe-sniff pppoe-sniff.o if.o common.o debug.o

 pppoe-server: pppoe-server.o if.o debug.o common.o md5.o libevent/libevent.a @PPPOE_SERVER_DEPS@
-	@CC@ -o pppoe-server @RDYNAMIC@ pppoe-server.o if.o debug.o common.o md5.o $(PPPOE_SERVER_LIBS) -Llibevent -levent
+	@CC@ $(LDFLAGS) -o pppoe-server @RDYNAMIC@ pppoe-server.o if.o debug.o common.o md5.o $(PPPOE_SERVER_LIBS) -Llibevent -levent

 pppoe: pppoe.o if.o debug.o common.o ppp.o discovery.o
-	@CC@ -o pppoe pppoe.o if.o debug.o common.o ppp.o discovery.o
+	@CC@ $(LDFLAGS) -o pppoe pppoe.o if.o debug.o common.o ppp.o discovery.o

 pppoe-relay: relay.o if.o debug.o common.o
-	@CC@ -o pppoe-relay relay.o if.o debug.o common.o
+	@CC@ $(LDFLAGS) -o pppoe-relay relay.o if.o debug.o common.o

 pppoe.o: pppoe.c pppoe.h
	@CC@ $(CFLAGS) '-DVERSION="$(VERSION)"' -c -o pppoe.o pppoe.c
@@ -113,7 +113,7 @@ relay.o: relay.c relay.h pppoe.h

 # Linux-specific plugin
 rp-pppoe.so: plugin/libplugin.a plugin/plugin.o
-	@CC@ -o rp-pppoe.so -shared plugin/plugin.o plugin/libplugin.a
+	@CC@ $(LDFLAGS) -o rp-pppoe.so -shared plugin/plugin.o plugin/libplugin.a

 plugin/plugin.o: plugin.c
	@CC@ '-DRP_VERSION="$(VERSION)"' $(CFLAGS) -I$(PPPD_INCDIR) -c -o plugin/plugin.o -fPIC plugin.c
